### PR TITLE
prw2.0: Added support for "custom" layouts for native histogram proto

### PR DESCRIPTION
**Pull request by bwplotka**
Result of the discussions:
* https://github.com/prometheus/prometheus/issues/13475#issuecomment-1931496924
* https://cloud-native.slack.com/archives/C02KR205UMU/p1707301006347199

cc @krajorama @beorn7 @zenador

### CODE REVIEW

File: prompb/write/v2/types.proto [hunk 1]
```diff
@@ -111,23 +113,43 @@ message Histogram {
...
   int64 timestamp = 15;
+
+  // custom_bounds specify monotonically increasing upper inclusive boundary for
+  // the bucket counts in positive_counts or/and negative_counts (or corresponding
+  // deltas) with arbitrary widths for this histogram. For each upper boundary
+  // the previous boundary represent lower exclusive/ boundary for that bucket.
+  //
+  // The first element is the upper inclusive boundary for the first bucket,
+  // which implicitly has a lower inclusive bound of -Inf. This is similar to
+  // "le" label semantics on classic histograms. You may add a bucket with an
```
Comment (krajorama) [inline prompb/write/v2/types.proto:141]: Not sure we want to get into the question of adding a 0 boundary, that's more a question of instrumentation/visualization, feels out of scope for storage.

File: prompb/write/v2/types.proto [hunk 2]
```diff
@@ -111,23 +113,43 @@ message Histogram {
   // Negative Buckets.
   repeated BucketSpan negative_spans =  8 [(gogoproto.nullable) = false];
   // Use either "negative_deltas" or "negative_counts", the former for
-  // regular histograms with integer counts, the latter for float
-  // histograms.
+  // regular exponential histograms with integer counts, the latter for exponential
```
Comment (krajorama) [inline prompb/write/v2/types.proto:116]: I don't think this is necessary.

File: prompb/write/v2/types.proto [hunk 3]
```diff
@@ -111,23 +113,43 @@ message Histogram {
...
+  // float histograms.
   repeated sint64 negative_deltas    =  9; // Count delta of each bucket compared to previous one (or to zero for 1st bucket).
   repeated double negative_counts    = 10; // Absolute count of each bucket.
 
   // Positive Buckets.
   repeated BucketSpan positive_spans = 11 [(gogoproto.nullable) = false];
   // Use either "positive_deltas" or "positive_counts", the former for
-  // regular histograms with integer counts, the latter for float
-  // histograms.
+  // regular exponential histograms with integer counts, the latter for exponential
```
Comment (krajorama) [inline prompb/write/v2/types.proto:124]: I don't think this is necessary.

File: prompb/write/v2/types.proto [hunk 4]
```diff
@@ -111,23 +113,43 @@ message Histogram {
...
   repeated BucketSpan negative_spans =  8 [(gogoproto.nullable) = false];
   // Use either "negative_deltas" or "negative_counts", the former for
-  // regular histograms with integer counts, the latter for float
-  // histograms.
+  // regular exponential histograms with integer counts, the latter for exponential
+  // float histograms.
   repeated sint64 negative_deltas    =  9; // Count delta of each bucket compared to previous one (or to zero for 1st bucket).
   repeated double negative_counts    = 10; // Absolute count of each bucket.
 
   // Positive Buckets.
```
Comment (krajorama) [inline prompb/write/v2/types.proto:121]:
```suggestion
  // Positive Buckets.
  // In case of the use of "custom_bounds" the positive buckets are interpreted as follows:
  // the span offset+length points to an the index of the custom_bound array or +Inf if indexing outside
  // the array. The counts and deltas have the same meaning as for exponential histograms.
```

File: prompb/write/v2/types.proto (hunk 2 above)
Comment (bwplotka) [reply prompb/write/v2/types.proto:116]: oh this is even wrong, sorry, I meant to do this on negative/positive spans. Should I?

File: prompb/write/v2/types.proto (hunk 4 above)
Comment (bwplotka) [reply prompb/write/v2/types.proto:121]: Or... I thought we ignore spans (why should we use them?)

File: prompb/write/v2/types.proto (hunk 1 above)
Comment (bwplotka) [reply prompb/write/v2/types.proto:141]: Sure, just I want users to not add 0 first bound because they think they have to

File: prompb/write/v2/types.proto (hunk 4 above)
Comment (krajorama) [reply prompb/write/v2/types.proto:121]: For code reuse and also because there are plenty of sparse classic histograms out there. For example when you application works smoothly , only the first 1-2 buckets would be used out of 10-40 possibly

File: prompb/write/v2/types.proto (hunk 2 above)
Comment (krajorama) [reply prompb/write/v2/types.proto:116]: it's not wrong, but I think it's better to explain the usage on the bucket level (where it says Negative Buckets, Positive buckets) because we might be reusing these later for other schemas.

File: prompb/write/v2/types.proto [hunk 5]
```diff
@@ -101,6 +101,8 @@ message Histogram {
   // bucket boundary is the previous boundary times 2^(2^-n). In the
   // future, more bucket schemas may be added using numbers < -4 or >
   // 8.
+  // TODO(bwplotka): Explain schema semantics for custom_bounds (custom histograms)
```
[the remark this answers is not in the document]
Comment (beorn7) [reply prompb/write/v2/types.proto:104]: I like 127. Where "schema 127" means: custom_bounds are used to specify explicit upper bounds.

**Review by npazosmendez — APPROVED** [review]
LGTM

Krajo's nits probably make sense but I don't have enough context

File: prompb/write/v2/types.proto [hunk 6]
```diff
@@ -111,23 +113,54 @@ message Histogram {
...
+  // boundary for that bucket. The first element is the upper inclusive boundary
+  // for the first bucket, which implicitly has a lower inclusive bound of -Inf.
+  // This is similar to "le" label semantics on classic histograms. You may add a
+  // bucket with an upper bound of 0 to make sure that you really have no negative
+  // observations, but in practice, native histogram rendering will show both with
+  // or without first upper boundary 0 and no negative counts as the same case.
+  //
+  // The last element is a lower bound for the implicit +Inf bucket (the overflow
+  // bucket).
+  repeated double custom_bounds = 16;
```
Comment (krajorama) [inline prompb/write/v2/types.proto:163]: @bwplotka . On the last histograms sync we realized that supporting other schemas that need extra values could be sooner rather than later. So we might want to actually have a more neutral name for this field like `custom_values`.

Comment (bwplotka) [reply prompb/write/v2/types.proto:163]: Got it, makes sense. But comments stays for now? Or should it `custom_values for schema 127 means custom bounds ...`?

File: prompb/write/v2/types.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (beorn7) [reply prompb/write/v2/types.proto:104]: We had one more fun idea about the schema: -53 (in hex: 0xCB, short for Custom Buckets). (We only need to reserve -10 and above for theoretical cases to cover "all of float64 in a single bucket" in the conventional exponential schemas.)

But if 127 is already used anywhere in checked-in code, let's not change it now just for the joke.

[the remark this answers is not in the document]
Comment (zenador) [reply prompb/write/v2/types.proto:104]: I don't think 127 is used in checked-in code yet, so up to you.

File: prompb/write/v2/types.proto (hunk 6 above)
Comment (bwplotka) [reply prompb/write/v2/types.proto:163]: I would say values

File: prompb/write/v2/types.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply prompb/write/v2/types.proto:104]: I like the -53 and -10, sounds fun

[the remark this answers is not in the document]
Comment (beorn7) [reply prompb/write/v2/types.proto:104]: OK, then let's take -53 for custom buckets.

File: prompb/write/v2/types.proto [hunk 7]
```diff
@@ -111,23 +113,54 @@ message Histogram {
...
-  // histograms.
+  // regular histograms with integer counts, the latter for
+  // float histograms.
   repeated sint64 negative_deltas    =  9; // Count delta of each bucket compared to previous one (or to zero for 1st bucket).
   repeated double negative_counts    = 10; // Absolute count of each bucket.
 
   // Positive Buckets.
+  // In case of "custom_bounds" use, the positive buckets are interpreted as follows:
+  // * The span offset+length points to an the index of the custom_bound array
+  // or +Inf if pointing to outside the array.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply prompb/write/v2/types.proto]: Thanks, clarified!

File: prompb/write/v2/types.proto (hunk 6 above)
Comment (bwplotka) [reply prompb/write/v2/types.proto]: Done!

File: prompb/write/v2/types.proto [hunk 8]
```diff
@@ -172,16 +177,21 @@ message Histogram {
...
+  // regular histograms with integer counts, the latter for
+  // float histograms.
   repeated sint64 negative_deltas    =  9; // Count delta of each bucket compared to previous one (or to zero for 1st bucket).
   repeated double negative_counts    = 10; // Absolute count of each bucket.
 
   // Positive Buckets.
+  //
+  // In case of custom buckets (-53 schema value) the positive buckets are interpreted as follows:
+  // * The span offset+length points to an the index of the custom_values array
+  // or +Inf if pointing to the len or array.
```
Comment (krajorama) [inline prompb/write/v2/types.proto:189]:
```suggestion
  // or +Inf if pointing to the len of array.
```

**Review by krajorama — APPROVED** [review]
Approved with nit on a typo

File: prompb/write/v2/types.proto [hunk 9]
```diff
@@ -190,6 +200,35 @@ message Histogram {
...
+  // For each upper boundary the previous boundary represent the lower exclusive
+  // boundary for that bucket. The first element is the upper inclusive boundary
+  // for the first bucket, which implicitly has a lower inclusive bound of -Inf.
+  // This is similar to "le" label semantics on classic histograms. You may add a
+  // bucket with an upper bound of 0 to make sure that you really have no negative
+  // observations, but in practice, native histogram rendering will show both with
+  // or without first upper boundary 0 and no negative counts as the same case.
+  //
+  // The last element is a lower bound for the implicit +Inf bucket (the overflow
+  // bucket).
```
Comment (beorn7) [inline prompb/write/v2/types.proto:230]: That's not consistent with the above (which says that the +Inf bucket is marked by pointing to len(custom_values)).

File: prompb/write/v2/types.proto [hunk 10]
```diff
@@ -190,6 +200,35 @@ message Histogram {
   // For Go, see github.com/prometheus/prometheus/model/timestamp/timestamp.go
   // for conversion from/to time.Time to Prometheus timestamp.
   int64 timestamp = 15;
+
+  // custom_values is an additional field used by non-exponential bucketing layouts.
+  //
+  // For custom buckets (-53 schema value) custom_values specify monotonically
+  // increasing upper inclusive boundary for the bucket counts with arbitrary
```
Comment (beorn7) [inline prompb/write/v2/types.proto:207]: boundary → boundaries

**Review by beorn7 — CHANGES_REQUESTED** [review]

File: prompb/write/v2/types.proto (hunk 9 above)
Comment (beorn7) [reply prompb/write/v2/types.proto:230]: OK, I take that back. It _is_ consistent, but it is confusing because that element is _also_ the upper bound of the last non-Inf bucket.

Maybe word that better. Suggestion: "The last element is not only the upper inclusive bound of the last regular bucket, but implicitly the lower exclusive bound of the +Inf bucket."

Comment (beorn7) [reply prompb/write/v2/types.proto:230]: Or remove this sentence entirely, as things are explained above for the +Inf bucket.

**Comment by bwplotka** [comment]
Addressed all and fixed tests. Thanks @beorn7 for patience!

**Review by beorn7 — APPROVED** [review]
Woah, the `OptimizedMarshalToSizedBuffer` scares me. But approving anyway, provided that tests have passed.

**Comment by bwplotka** [comment]
Yea, this might get replaced with new protobuf (+gogo removal)
